core/src/main/kotlin/discord4j/core/object/entity/MessageChannelExtensions.kt

```diff
@@ -0,0 +1,28 @@
+package discord4j.core.`object`.entity
+
+import discord4j.core.*
+import discord4j.core.`object`.util.Snowflake
+import discord4j.core.spec.EmbedCreateSpec
+import discord4j.core.spec.MessageCreateSpec
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.reactive.publish
+import java.time.Instant
+
+fun MessageChannel.lastMessageId(): Snowflake? = lastMessageId.grab()
+suspend fun MessageChannel.lastMessage(): Message? = lastMessage.awaitNull()
+fun MessageChannel.lastPinTimestamp(): Instant? = lastPinTimestamp.grab()
+suspend fun MessageChannel.newMessage(spec: (MessageCreateSpec) -> Unit): Message = createMessage(spec).await()
+suspend fun MessageChannel.newMessage(message: String): Message = newMessage { it.setContent(message) }
+suspend fun MessageChannel.newEmbed(spec: (EmbedCreateSpec) -> Unit): Message = newMessage { it.setEmbed(spec) }
+suspend fun MessageChannel.awaitType(): Unit = type().unit()
+suspend fun MessageChannel.awaitTypeUntil(trigger: suspend ProducerScope<*>.() -> Unit): ReceiveChannel<Long> =
```

Comment:
So this actually just straight up won't work. I know I told you to do this, but I didn't know coroutines as well as I've learned over the past few days.

So what actually happens here is `coroutineScope` won't ever resume until the body finishes...which won't ever happen actually because there would be no way to drain the channel while the coroutine has been suspended.

So here's how to fix this:
```
suspend fun MessageChannel.awaitTypeUntil(trigger: suspend () -> Unit): ReceiveChannel<Long> = suspendCoroutine {
    it.resume(typeUntil(CoroutineScope(it.context).publish { trigger() }).openSubscription())
}
```
This will create a scope and transform it into a Publisher, which when the trigger completes is when `onCompelete` is invoked, which will stop the typing.